secboot/tpm.go

```diff
@@ -0,0 +1,209 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"crypto/rand"
+	"fmt"
+
+	"github.com/chrisccoulson/go-tpm2"
+	sb "github.com/snapcore/secboot"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+const (
+	// Handles are in the block reserved for owner objects (0x01800000 - 0x01bfffff)
+	pinHandle = 0x01800000
+)
+
+type tpmSupport struct {
+	// Connection to the TPM device
+	tconn *sb.TPMConnection
+	// Lockout authorization
+	lockoutAuth []byte
+	// Paths to shim files
+	shimFiles []string
+	// Paths to bootloader files
+	bootloaderFiles []string
+	// Paths to kernel files
+	kernelFiles []string
+	// Kernel command lines
+	kernelCmdlines []string
+}
+
+func NewTPMSupport() (*tpmSupport, error) {
+	tconn, err := sb.ConnectToDefaultTPM()
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to default TPM: %v", err)
+	}
+
+	lockoutAuth := make([]byte, 16)
+	// crypto rand is protected against short reads
+	_, err = rand.Read(lockoutAuth)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create lockout authorization: %v", err)
+	}
+
+	t := &tpmSupport{tconn: tconn, lockoutAuth: lockoutAuth}
+
+	return t, nil
+}
+
+// StoreLockoutAuth saves the lockout authorization data in a file at the
+// path specified by filename.
+func (t *tpmSupport) StoreLockoutAuth(filename string) error {
+	if err := osutil.AtomicWriteFile(filename, t.lockoutAuth, 0600, 0); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetShimFiles verifies and sets the list of shim binaries.
+func (t *tpmSupport) SetShimFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.shimFiles = pathList
+	return nil
+}
+
+// SetBootloaderFiles verifies and sets the list of bootloader binaries.
+func (t *tpmSupport) SetBootloaderFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.bootloaderFiles = pathList
+	return nil
+}
+
+// SetKernelFiles verifies and sets the list of kernel binaries.
+func (t *tpmSupport) SetKernelFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.kernelFiles = pathList
+	return nil
+}
+
+func ensureFilesExist(pathList []string) error {
+	for _, p := range pathList {
+		if !osutil.FileExists(p) {
+			return fmt.Errorf("file %s does not exist", p)
+		}
+	}
+	return nil
+}
+
+var (
+	sbProvisionTPM = sb.ProvisionTPM
+)
+
+// Provision tries to clear and provision the TPM.
+func (t *tpmSupport) Provision() error {
+	if err := sbProvisionTPM(t.tconn, sb.ProvisionModeFull, t.lockoutAuth); err != nil {
```

Comment:
Contrary to the comment, this doesn't clear the TPM - it performs all actions without clearing it first. As a consequence, this will fail on a device that has previously been provisioned because the current lockout hierarchy authorization value for the TPM will be incorrect (it's unknown at this point).

Reply:
Fixing the comment.